it/tutorials/scala-for-java-programmers.md

```diff
@@ -0,0 +1,716 @@
+-   le funzioni getter sono automaticamente definite per i parametri del
+    costruttore (i.e. è possibile ricavare il valore del parametro `v` del
+    costruttore di qualche istanza della classe `c` semplicemente
+    scrivendo `c.v`),
+-   sono disponibili le definizioni di default dei metodi `equals` e
+    `hashCode` che lavorano sulle *strutture* delle istanze e non sulle
+    loro identità,
+-   è disponibile la definizione di default del metodo `toString` che stampa
+    il valore in “source form” (e.g. l’albero per l’espressione `x+1` stampa
+    `Sum(Var(x),Const(1))`),
+-   le istanze di queste classi possono essere decomposte con il
+    *pattern matching* come vedremo più avanti.
+
+Ora che abbiamo definito il tipo dati per rappresentare le nostre
+espressioni aritmetiche possiamo iniziare a definire le operazioni per
+manipolarle. Iniziamo con una funzione per valutare l’espressione in un
+qualche *ambiente* (_environment_) di valutazione. Lo scopo dell’environment
+è quello di dare i valori alle variabili. Per esempio, l’espressione
+`x+1` valutata nell’environment con associato il valore `5` alla
+variabile `x`, scritto `{ x -> 5 }`, restituisce `6` come risultato.
+
+Inoltre, dobbiamo trovare un modo per rappresentare gli environment.
+Potremmo naturalmente usare alcune strutture dati associative come una
+hash table ma, possiamo anche usare direttamente delle funzioni! Un
+environment in realtà non è altro che una funzione con associato un
+valore al nome di una variabile. L’environment `{ x -> 5 }`
+mostrato sopra può essere semplicemente scritto in Scala come:
+
+    { case "x" => 5 }
+
+Questa notazione definisce una funzione che quando riceve la stringa `"x"`
+come argomento restituisce l’intero `5` e fallisce con un’eccezione negli
+altri casi.
+
+Prima di scrivere la funzione di valutazione diamo un nome al tipo di
+environment. Potremmo usare sempre il tipo `String => Int` per gli environment
+ma semplifichiamo il programma se introduciamo un nome per questo tipo
+rendendo anche i cambiamenti futuri più facili. Questo è fatto in con la
+seguente notazione:
+
+    type Environment = String => Int
+
+Da ora in avanti il tipo `Environment` può essere usato come un alias per
+il tipo delle funzioni da `String` a `Int`.
+
+Possiamo ora passare alla definizione della funzione di valutazione.
+Concettualmente è molto semplice: il valore della somma di due
+espressioni è pari alla somma dei valori delle loro espressioni; il
+valore di una variabile è ottenuto direttamente dall’environment; il
+valore di una costante è la costante stessa. Esprimere quanto appena
+detto in Scala non è difficile:
+
+    def eval(t: Tree, env: Environment): Int = t match {
+      case Sum(l, r) => eval(l, env) + eval(r, env)
+      case Var(n)    => env(n)
+      case Const(v)  => v
+    }
+
+Questa funzione di valutazione lavora effettuando un *pattern matching*
+sull’albero `t`. Intuitivamente il significato della definizione precendente
```

Comment:
Sine you have made the parallel with Java often, maybe it's worth pointing out that the Visitor pattern is nothing more than an elaborated way of implementing pattern matching in a language that doesn't offer native support for it.